lib/quicly.c

```diff
@@ -4333,9 +4350,20 @@ static int do_send(quicly_conn_t *conn, quicly_send_context_t *s)
     /* handle timeouts */
     if (conn->idle_timeout.at <= conn->stash.now) {
         QUICLY_PROBE(IDLE_TIMEOUT, conn, conn->stash.now);
-        conn->super.state = QUICLY_STATE_DRAINING;
-        destroy_all_streams(conn, 0, 0);
-        return QUICLY_ERROR_FREE_CONNECTION;
+        goto CloseNow;
+    }
+    /* handle handshake timeouts */
+    if ((conn->initial != NULL || conn->handshake != NULL) &&
+        conn->created_at + (uint64_t)conn->super.ctx->handshake_timeout_rtt_multiplier * conn->egress.loss.rtt.smoothed <=
+            conn->stash.now) {
+        QUICLY_PROBE(HANDSHAKE_TIMEOUT, conn, conn->stash.now, conn->stash.now - conn->created_at, conn->egress.loss.rtt.smoothed);
+        conn->super.stats.num_handshake_timeouts++;
+        goto CloseNow;
+    }
+    if (conn->super.stats.num_packets.initial_handshake_sent >= conn->super.ctx->max_initial_handshake_packets) {
+        QUICLY_PROBE(INITIAL_HANDSHAKE_PACKET_EXCEED, conn, conn->stash.now, conn->super.stats.num_packets.initial_handshake_sent);
+        conn->super.stats.num_initial_handshake_exceeded++;
```

Comment:
Do we have a chance of this value becoming other than 0 or 1? It seems to me that this is a boolean value rather than a counter.

Reply:
Hmm, that's a valid question. However, considering that we'll use this value for aggregation in h2o, it'd be natural to keep this as an integer rather than changing it to a boolean.

How about changing this line to
```C
conn->super.stats.num_initial_handshake_exceeded = 1;
```

Reply:
That's fair. I'm not sure how long we can continue reusing the same type for representing aggregated information, but I'm fine with going with what we have now (i.e., `conn->super.stats.num_initial_handshake_exceeded++;`).